File: baekjoonn/11053_가장 긴 증가하는 부분수열 길이/B_11053_2.js
const binarySearch = function(start, end, item) {
    if (start >= end) {
        return end
    }

    let mid = Math.floor((start + end) / 2)
    if (dp[mid] > item) {
        return binarySearch(start, mid, item)
    } else {
        if (dp[mid] === item) {
            return mid
        } else {
            return binarySearch(mid + 1, end, item)
        }
    }
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = Number(input[0])
const sequence = input[1].split(' ').map(v => Number(v))

let dp = [sequence[0]]
for (let i = 1; i < N; i++) {
    const target = sequence[i]
    if (dp[dp.length - 1] < target) {
        dp.push(target)
    } else {
        // 같은 경우는 제외
        if (dp[dp.length - 1] > target) {
            const idx = binarySearch(0, dp.length - 1, target)
            dp[idx] = target
        }
    }
}

console.log(dp.length)

File: baekjoonn/11053_가장 긴 증가하는 부분수열 길이/B_11053_2.py
N = int(input())
numbers = list(map(int, input().split()))
dp = [1] * N
max_dp = 1
if N > 1:
    for i in range(1, N):
        for j in range(i - 1, -1, -1):
            if numbers[i] > numbers[j]:
                dp[i] = max(dp[i], dp[j] + 1)

        if dp[i] > max_dp:
            max_dp = dp[i]

print(max_dp)
